### PR TITLE
Added tests for $where

### PR DESCRIPTION
**Pull request by thecalvinchan**
## Compares performance between queries using the query language and the $where operator.

- Comparing a field in a document to a static integer
- $in
- $elemMatch
- Regex
- Querying nested field, match to a static integer

## Checks performance for more complicated queries using $where

- Comparison between two fields in the same document
- Mixed query using query language and $where
- Comparison between two nested fields in the same document

## Document Shape and Size Experimentation

- Query a collection of very large documents with deeply nested fields using both, query language and $where

### CODE REVIEW

**Comment by amidvidy** [comment]
can you squash all these commits in to one commit?

File: testcases/where.js [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+                      collection.insert( { x : i } );
+                  }
+              },
+              ops : [
+                { op : "find", query : { x : 1 } }
+              ] } );
+
+/**
+ * Setup: creates a collection with 4800 documents of the form { x : i }
+ * Test: Finds { x : 1 } using a $where query with ===
```
Comment (amidvidy) [inline testcases/where.js]: elaborate that '===' is type-safe equality

File: testcases/where.js [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+                      collection.insert( { x : i } );
+                  }
+              },
+              ops : [
+                { op : "find", query : { $where : "this.x === 1" } } 
+              ] } );
+
+/**
+ * Setup: creates a collection with 4800 documents of the form { x : i }
+ * Test: Finds { x : 1 } using a $where query with ==
```
Comment (amidvidy) [inline testcases/where.js]: elaborate that '==' is weak equality

File: testcases/where.js [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+
+/**
+ * Setup: creates a collection with 4800 documents of the form { x : i }
+ * Test: Finds all documents with x between 1 and 6 using $in 
+ */
+tests.push( { name : "Where.v1.In.QueryLanguage",
+              tags: ['query','querylanguage'],
+              pre: function( collection ) {
+                  collection.drop();
+                  for ( var i = 0; i < 4800; i++ ) {
```
Comment (amidvidy) [inline testcases/where.js]: be consistent about coding style e.g.
` for ( var i = 0; i < 4800; i++ ) {`

vs

`for (var i = 0; i < 4800; i++) {`

I prefer the later.

File: testcases/where.js [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+              pre: function( collection ) {
+                  collection.drop();
+                  for ( var i = 0; i < 200; i++ ) {
+                    for ( var j = 0; j < 200; j++ ) {
+                      collection.insert( { x : i, y : j } );
+                    }
+                  }
+              },
+              ops : [
+                { op : "find", query : { $where : "this.x == this.y" } }
```
Comment (amidvidy) [inline testcases/where.js]: be consistent about $where syntax as well - e.g. passing the function as a string vs as an actual function. I prefer the latter.

File: testcases/where.js [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+ * random numbers between 0 and 100
+ * Test: Finds all documents with x containing y such that 80 <= y < 85 
+ * using $elemMatch
+ */
+tests.push( { name : "Where.v1.ElemMatch.QueryLanguage",
+              tags: ['query','querylanguage'],
+              pre: function( collection ) {
+                  collection.drop();
+                  for ( var i = 0; i < 4800; i++ ) {
+                    var numbers = [];
```
Comment (amidvidy) [inline testcases/where.js]: can you factor out the pre: block from here and ElemMatch.Where into a common utility function?

File: testcases/where.js [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+ * Setup: Create collection with documents containing 4 character alphabetic permutations
+ * Test: Find document based on Regex Query
+ */
+tests.push( { name : "Where.v1.Regex.QueryLanguage",
+            tags: ['query','querylanguage'],
+            pre: function(collection) {
+                collection.drop();
+                var coll = generatePermutations();
+                collection.insert(coll)
+            },
```
Comment (amidvidy) [inline testcases/where.js]: same about refactoring the pre blocks

File: testcases/where.js [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+
+/*
+ * Setup: Creates a collection of 13 objects, each with 4 nested levels of 13 fields
+ * Test: Find document through match of deeply nested field using $where
+ */
+tests.push( { name : "Where.v1.SimpleNested.Where",
+            tags: ['query', 'where'],
+            pre: function(collection) {
+                collection.drop();
+                var coll = generateNestedCollection();
```
Comment (amidvidy) [inline testcases/where.js]: same about refactoring the pre blocks

**Comment by dalyd** [comment]
Could you post using the code review tool? https://wiki.mongodb.com/display/KERNEL/How+We+Use+Code+Review

File: testcases/where.js [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+    tests = [];
+}
+
+// Queries that can be written in query language and using $where
+
+/**
+ * Setup: creates a collection with 4800 documents of the form { x : i }
+ * Test: Finds { x : 1 } using query language
+ */
+tests.push( { name : "Where.v1.CompareToInt.QueryLanguage",
```
Comment (dalyd) [inline testcases/where.js]: Please remove the .v1 in the testnames. We used to have our tests versioned with v1, v2, etc, but we've taken that out of the test files.

**Pull request closed without merging**